### PR TITLE
Feature/deployment stragegy

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Add update strategy for stateful|stateless|daemon applications;

## QA steps

```console
## stateful
$ yml2json /tmp/charm-builds/mariadb-k8s/metadata.yaml | jq .deployment
{
  "type": "stateful",
  "min-version": "1.10.1",
  "service": "omit"
}

$ juju run --unit mariadb-k8s/0  pod-spec-get | yml2json | jq .service
{
  "updateStrategy": {
    "rollingUpdate": {
      "partition": 10
    },
    "type": "RollingUpdate"
  }
}

$ mkubectl get -nt1 statefulset.apps/mariadb-k8s -o json | jq .spec.updateStrategy
{
  "rollingUpdate": {
    "partition": 10
  },
  "type": "RollingUpdate"
}

## stateless
$ yml2json /tmp/charm-builds/mariadb-k8s/metadata.yaml | jq .deployment
{
  "type": "stateless",
  "min-version": "1.10.1",
  "service": "omit"
}

$ juju run --unit mariadb-k8s/0  pod-spec-get | yml2json | jq .service
{
  "updateStrategy": {
    "rollingUpdate": {
      "maxUnavailable": 10
    },
    "type": "RollingUpdate"
  }
}

$ mkubectl get -nt2 deployment.apps/mariadb-k8s -o json | jq .spec.strategy
{
  "rollingUpdate": {
    "maxSurge": "25%",
    "maxUnavailable": 10
  },
  "type": "RollingUpdate"
}

## daemon
$ yml2json /tmp/charm-builds/mariadb-k8s/metadata.yaml | jq .deployment
{
  "type": "daemon",
  "min-version": "1.10.1",
  "service": "omit"
}

$ juju run --unit mariadb-k8s/0  pod-spec-get | yml2json | jq .service
{
  "updateStrategy": {
    "rollingUpdate": {
      "maxUnavailable": 10
    },
    "type": "RollingUpdate"
  }
}

$ mkubectl get -nt4 daemonset.apps/mariadb-k8s -o json | jq .spec.updateStrategy
{
  "rollingUpdate": {
    "maxUnavailable": 10
  },
  "type": "RollingUpdate"
}

```

## Documentation changes

Yes

## Bug reference

https://bugs.launchpad.net/juju/+bug/1891206

### CODE REVIEW

**Comment by ycliuhw** [comment]
!!build!!

File: caas/kubernetes/provider/specs/types.go [hunk 1]
```diff
@@ -205,3 +206,16 @@ func getParser(specVersion specs.Version) (parserType, error) {
 		return nil, errors.NewNotSupported(nil, fmt.Sprintf("latest supported version %d, but got podspec version %d", specs.CurrentVersion, specVersion))
 	}
 }
+
+// IntOrStringToK8s converts IntOrString to k8s version.
+func IntOrStringToK8s(in specs.IntOrString) (*intstr.IntOrString, error) {
+	data, err := in.MarshalJSON()
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go:212]: I wonder why we can't just copy the fields into a result struct?
Less expensive than marshalling?

Comment (ycliuhw) [reply caas/kubernetes/provider/specs/types.go]: right

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by manadart** [comment]
@ycliuhw @wallyworld We're going to have to be on top of these doc changes. Is there a card/ticket to incorporate this change?

**Pull request merged**